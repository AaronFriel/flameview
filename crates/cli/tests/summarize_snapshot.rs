#![cfg(not(miri))]

use std::path::{Path, PathBuf};
use std::process::Command;

use regex::Regex;

fn run_summary(fixture: &Path, max_lines: usize, coverage: f64) -> String {
    let bin = env!("CARGO_BIN_EXE_flameview");
    let output = Command::new(bin)
        .arg(fixture)
        .arg("--summarize")
        .arg("--max-lines")
        .arg(max_lines.to_string())
        .arg("--coverage")
        .arg(coverage.to_string())
        .output()
        .expect("run");
    assert!(
        output.status.success(),
        "{}",
        String::from_utf8_lossy(&output.stderr)
    );
    let stdout = String::from_utf8(output.stdout).expect("utf8");
    let re = Regex::new(r"[A-Z]:[/\\][^\s]*").expect("regex");
    let cleaned = re.replace_all(&stdout, "");
    let cleaned = cleaned.replace("\r\n", "\n");
    let lines: Vec<_> = cleaned.lines().map(|l| l.trim_end()).collect();
    lines.join("\n").trim_end_matches('\n').to_string()
}

#[test]
fn summarize_snapshots() {
    let data_dir: PathBuf = [env!("CARGO_MANIFEST_DIR"), "..", "..", "tests", "data"]
        .into_iter()
        .collect();

    let cases = [
        ("small_default", data_dir.join("small.txt"), 50, 0.95),
        ("medium_30_90", data_dir.join("medium.txt"), 30, 0.90),
        ("large_15_80", data_dir.join("large.txt"), 15, 0.80),
    ];

    for (name, fixture, lines, cov) in cases {
        let output = run_summary(&fixture, lines, cov);
        let mut settings = insta::Settings::new();
        settings.set_prepend_module_to_snapshot(false);
        settings.bind(|| insta::assert_snapshot!(name, output));
    }
}

#[test]
#[should_panic]
fn summarize_corrupt_panics() {
    let bad: PathBuf = [env!("CARGO_MANIFEST_DIR"), "..", "..", "README.md"]
        .into_iter()
        .collect();

    run_summary(&bad, 10, 0.5);
}
