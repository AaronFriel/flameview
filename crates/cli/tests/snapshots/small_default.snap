---
source: crates/cli/tests/summarize_snapshot.rs
expression: output
---
(root) (100.0%, 34)
  java (79.4%, 27)
    start_thread (79.4%, 27)
      java_start (79.4%, 27)
        JavaThread::run (79.4%, 27)
          JavaThread::thread_main_inner (79.4%, 27)
            thread_entry (79.4%, 27)
              JavaCalls::call_virtual (79.4%, 27)
                JavaCalls::call_virtual (79.4%, 27)
                  JavaCalls::call_helper (79.4%, 27)
                    call_stub_[j] (79.4%, 27)
                      Interpreter_[j] (79.4%, 27)
                        Interpreter_[j] (79.4%, 27)
                          io/netty/channel/nio/NioEventLoop:.run_[j] (79.4%, 27)
                            io/netty/channel/nio/NioEventLoop:.processSelectedKeysOptimized_[j] (79.4%, 27)
                              io/netty/channel/nio/NioEventLoop:.processSelectedKey_[j] (79.4%, 27)
                                io/netty/channel/nio/AbstractNioByteChannel$NioByteUnsafe:.read_[j] (79.4%, 27)
                                  io/netty/channel/DefaultChannelHandlerContext:.fireChannelRead_[j] (64.7%, 22)
                                    io/netty/handler/codec/ByteToMessageDecoder:.channelRead_[j] (64.7%, 22)
                                      io/netty/channel/DefaultChannelHandlerContext:.fireChannelRead_[j] (64.7%, 22)
                                        org/vertx/java/core/net/impl/VertxHandler:.channelRead_[j] (61.8%, 21)
                                          org/vertx/java/core/http/impl/VertxHttpHandler:.channelRead_[j] (61.8%, 21)
                                            org/vertx/java/core/http/impl/DefaultHttpServer$ServerHandler:.doMessageReceived_[j] (61.8%, 21)
                                              org/vertx/java/core/http/impl/ServerConnection:.handleMessage_[j] (61.8%, 21)
                                                org/mozilla/javascript/gen/file__home_bgregg_vert_x_2_1_sys_mods_io_vertx_lang_js_1_1_0_vertx_http_js_93:.call_[j] (20.6%, 7)
                                                  org/mozilla/javascript/gen/file__home_bgregg_vert_x_2_1_sys_mods_io_vertx_lang_js_1_1_0_vertx_http_js_93:.call_[j] (20.6%, 7)
                                                    org/mozilla/javascript/BaseFunction:.construct_[j] (17.6%, 6)
                                                      org/mozilla/javascript/gen/file__home_bgregg_vert_x_2_1_sys_mods_io_vertx_lang_js_1_1_0_vertx_http_js_93:.call_[j] (17.6%, 6)
                                                        org/mozilla/javascript/gen/file__home_bgregg_vert_x_2_1_sys_mods_io_vertx_lang_js_1_1_0_vertx_http_js_93:._c_anonymous_3_[j] (17.6%, 6)
                                                          org/mozilla/javascript/gen/file__home_bgregg_vert_x_2_1_sys_mods_io_vertx_lang_js_1_1_0_vertx_http_js_93:.<init>_[j] (11.8%, 4)
                                                            org/mozilla/javascript/NativeFunction:.initScriptFunction_[j] (11.8%, 4)
                                                              org/mozilla/javascript/TopLevel:.getBuiltinPrototype_[j] (11.8%, 4)
                                                                call_function_single_interrupt_[k] (11.8%, 4)
                                                                  smp_call_function_single_interrupt_[k] (11.8%, 4)
                                                                    generic_smp_call_function_single_interrupt_[k] (11.8%, 4)
                                                                      remote_function_[k] (11.8%, 4)
                                                                        __perf_event_enable_[k] (11.8%, 4)
                                                                          group_sched_in_[k] (11.8%, 4)
                                                                            x86_pmu_commit_txn_[k] (11.8%, 4)
                                                                              perf_pmu_enable_[k] (11.8%, 4)
                                                                                x86_pmu_enable_[k] (11.8%, 4)
                                                                                  intel_pmu_enable_all_[k] (11.8%, 4)
                                                                                    native_write_msr_safe_[k] (11.8%, 4)
                                                          org/mozilla/javascript/ScriptRuntime:.setObjectProp_[j] (2.9%, 1)
                                                            org/mozilla/javascript/IdScriptableObject:.has_[j] (2.9%, 1)
                                                              org/mozilla/javascript/ScriptableObject:.getSlot_[j] (2.9%, 1)
                                                          org/mozilla/javascript/BaseFunction:.construct_[j] (2.9%, 1)
                                                            org/mozilla/javascript/gen/file__home_bgregg_vert_x_2_1_sys_mods_io_vertx_lang_js_1_1_0_vertx_http_js_93:.call_[j] (2.9%, 1)
                                                              org/mozilla/javascript/gen/file__home_bgregg_vert_x_2_1_sys_mods_io_vertx_lang_js_1_1_0_vertx_http_js_93:._c_anonymous_21_[j] (2.9%, 1)
                                                                org/mozilla/javascript/gen/file__home_bgregg_vert_x_2_1_sys_mods_io_vertx_lang_js_1_1_0_vertx_http_js_93:.getParamAndVarCount_[j] (2.9%, 1)
                                                org/mozilla/javascript/gen/file__home_bgregg_vert_x_2_1_sys_mods_io_vertx_lang_js_1_1_0_vertx_http_js_91:.call_[j] (11.8%, 4)
